GS-640 Bulk Upload of FacetoFace sessions

Sajonasj-OC opened the pull request:
## Description:
<!-- Describe your changes in detail / numbered list of distinct changes for reference -->
## Pull Request Summary: Bulk Session Upload for Face-to-Face Activities

This pull request adds a feature that allows the bulk upload of sessions for face-to-face activities. The new functionality inserts session details into both the `facetoface` and `facetoface_dates` databases. An `example_bulk.csv` file is provided for reference.

### Key Features

- **Bulk Upload:** Enables the upload of multiple sessions simultaneously via CSV.
- **Database Integration:** Automatically inserts session data into the relevant databases.
- **Example CSV:** Provides an `example_bulk.csv` file to illustrate the expected format.

### Validation Rules

The following table outlines the validation rules enforced during the bulk upload process:

| **Field**                    | **Validation Rule**                                         |
|------------------------------|-------------------------------------------------------------|
| **Start Date & Time**        | Must exist and be a valid timestamp.                        |
| **Finish Date & Time**       | Must exist and be a valid timestamp.                        |
| **Start Time < Finish Time** | The start time must occur before the finish time.           |
| **Capacity**                 | Must be a positive integer.                                 |
| **Duration**                 | Must be a positive integer.                                 |
| **Normal Cost (Optional)**   | If provided, must be a valid number.                        |
| **Discount Cost (Optional)** | If provided, must be a valid number.                        |
| **Allow Cancellations**      | Must be either "yes" or "no".                                |
| **Allow Overbookings**       | Must be either "yes" or "no".                                |

### Error Handling

- If any errors are encountered during CSV processing, they will be displayed on the confirmation page.
- Error messages will clearly indicate which validations have failed, allowing users to correct the CSV data accordingly.

### Confirmation Page

- The confirmation page will display the data that is about to be inserted into the databases.
- This final review step allows users to verify the upload details before committing the data.
- Any validation errors detected during CSV processing will also be shown on this page to aid in troubleshooting.

---

### Summary of Added and Updated Files

**Added Files:**
- `bulk_session_manager.php`
- `bulk_session_confirm_form.php`
- `bulk_session_upload_form.php`
- `example_bulk.csv`
- `uploadbulksessions.php`
- 'csv_processed_bulksession.php'

**Updated Files:**
- `access.php`
- `facetoface.php`
- `view.php`

---

## Checklist before requesting a review:
- [x] I have performed a self review of my code.
- [x] My code follows the [Moodle Coding Style](https://moodledev.io/general/development/policies/codingstyle).
- [x] Version numbers have been updated.
- [x] Code has been commented, particularly in hard-to-understand areas.

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -0,0 +1,367 @@
+ *
+ * @author Jonas Sajonas
+ * @copyright GCHLOL, 2025
+ * @license http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
```
I'll get you to update this to the format we settled on. Copied in below for reference:
```php
/**
 * $DESCRIPTION$.
 *
 * @package     $PACKAGE$
 * @copyright   $YEAR$ Gold Coast Health
 * @author      Nicholas Lambell
 * @license     https://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
 */
```

NickLambellGCH commented on classes/form/bulk_session_confirm_form.php:
```diff
@@ -0,0 +1,63 @@
+ *
+ * @author    Jonas Sajonas
+ * @copyright GCHLOL, 2025
+ * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
```
This can be updated too

NickLambellGCH commented on classes/form/bulk_session_upload_form.php:
```diff
@@ -0,0 +1,107 @@
+ *
+ * @author     Jonas Sajonas
+ * @copyright  GCHLOL, 2025
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
```
This can be updated too

NickLambellGCH commented on uploadbulksessions.php:
```diff
@@ -0,0 +1,173 @@
+ *
+ * @author     Jonas Sajonas
+ * @copyright  GCHLOL, 2025
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
```
This can be updated too

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -0,0 +1,367 @@
+
+    /** Avoid undefined properties */
+    private $usefile;
+    private $file;
```
You may have added these to resolve 'undefined property' errors, but they shouldn't be defined / documented that way but rather just as any other property would be.

NickLambellGCH replied:
It would be good to see the properties have types since that's supported in the version of PHP that we target. e.g. `private int $facetofaceid`.

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -0,0 +1,367 @@
+    private $usefile;
+    private $file;
+
+    /** @var bool Will ignore case when matching users */
+    private $caseinsensitive = false;
+
+    /** @var bool When true, confirmation emails are not sent. */
+    private $suppressemail = false;
```
It looks like you're not using these properties other than setting them in `suppress_email()` and `set_case_insensitive()`.

Sajonasj-OC replied:
removed the properties as the bulk upload sessions is creating sessions and not signups.

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -0,0 +1,367 @@
+     * Load CSV data from a file (given its fileid).
+     *
+     * @param int $fileid
+     *
```
Typically we'll just group the `@param` and `@return` with no blank lines.

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -0,0 +1,367 @@
+    /**
+     * Load CSV data from a file (given its fileid).
+     *
+     * @param int $fileid
```
`@param` should have a description

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -0,0 +1,367 @@
+     *
+     * @return bool true if loaded successfully, false otherwise
+     */
+    public function load_from_file(int $fileid) {
```
Functions should optimally always have a return type defined, even if it's `void`.
e.g. `... load_from_file(...): bool {`

NickLambellGCH replied:
Looks like this wasn't addressed

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -0,0 +1,367 @@
+     * @return bool true if loaded successfully, false otherwise
+     */
+    public function load_from_file(int $fileid) {
+        global $USER;
```
It's good to have a blank line after the `global` line for readability.

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -0,0 +1,367 @@
+        $this->usefile = true;
+
+        $fs = get_file_storage();
+        $usercontext = \context_user::instance($USER->id);
```
Should import `context_user`

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -0,0 +1,367 @@
+        if (count($files) != 1) {
+            throw new moodle_exception('error:cannotloadfile', 'mod_facetoface');
+        }
+        $this->file = current($files);
```
Blank lines after closing braces are good for readability. The only real exception to this rule is usually closing resources that you're iterating on.
The main example of this is cases where you use `$DB->get_recordset_sql()`. In those cases, you'll usually see something like below.
```php
$recordset = $DB->get_recordset_sql(...);
foreach ($recordset as $record) {
    ...
}
$recordset->close();
```

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -0,0 +1,367 @@
+        $this->records = $records;
+
+        return $this;
+    }
```
Looks like this function is unused

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -0,0 +1,367 @@
+     *
+     * @return array
+     */
+    public static function get_headers() {
```
Function should have a return type

NickLambellGCH replied:
Looks like this wasn't addressed

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -0,0 +1,367 @@
+     *
+     * @return Generator
+     */
+    private function get_iterator(): \Generator {
```
`Generator` is already imported so doesn't need the `\` qualifier

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -0,0 +1,367 @@
+            foreach ($this->records as $record) {
+                yield $record;
+            }
+            return;
```
New line before `return` is good for readability

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -0,0 +1,367 @@
+                foreach ($data as &$field) {
+                    $field = trim($field, "'");
+                }
+                yield array_combine($headers, $data);
```
New line before key words is good. In this case both `foreach` and `yield`

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -0,0 +1,367 @@
+
+    /**
+     * Validate loaded CSV records.
+     * @return array an array of error messages (empty if no errors)
```
Should be a new line between the description and `@return` line

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -0,0 +1,367 @@
+     * Validate loaded CSV records.
+     * @return array an array of error messages (empty if no errors)
+     */
+    public function validate() {
```
Function should have return type

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -0,0 +1,367 @@
+            // Required: Allow Overbookings (must be "yes" or "no").
+            if (!in_array(strtolower($record['Allow overbookings'] ?? ''), ['yes', 'no'], true)) {
+                $this->errors[] = [$index, get_string('error:invalidallowoverbook', 'facetoface')];
+            }
```
Just a few general points for the cleanup of this section:

1. `if` statements with more than one condition should be expanded to have each condition on its own line. These should be separate from the parenthesis. e.g.
    ```php
    if (
        ... ||
        ...
    ) {
    ```
2. I try to avoid using `else` wherever possible as there's usually better ways to control flow but if used, there should always be a new line before the `else` line for readability. e.g.
    ```php
    if (...) {
        ...

    } else {
    ```

Sajonasj-OC replied:
refactored. I also hate too many nestings if statements

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -0,0 +1,367 @@
+            'Duration',
+            'Normal Cost',
+            'Discount Cost',
+            'Details',
```
I think a bit more consistency with the capitalisation of the headings would be good as some are all lower case, some only first word capitalised, and others all words capitalised.

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -0,0 +1,367 @@
+                $this->errors[] = [$index, get_string('error:invalidallowoverbook', 'facetoface')];
+            }
+        }
+        return $this->errors;
```
Should be a new line before the `return`.

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -0,0 +1,367 @@
+        global $DB;
+
+        foreach ($this->records as $record) {
+            $session = new \stdClass();
```
Should import `stdClass`.

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -0,0 +1,367 @@
+            // Combine Finish Date and Time.
+            $session->finishtime = (!empty($record['Finish date']) && !empty($record['Finish time']))
+            ? strtotime(str_replace('/', '-', $record['Finish date'] . ' ' . $record['Finish time']))
+            : null;
```
When there's this much content in a ternary, especially when the default is a simple value, I'd much prefer to see them expanded out to full `if` blocks for readability. There's a couple of these in this section.
e.g.
```php
$session->starttime = null;
if (
    !empty($record['Start date']) &&
    !empty($record['Start time'])
) {
    $session->starttime = strtotime(str_replace('/', '-', $record['Start date'] . ' ' . $record['Start time']));
} 
```

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -0,0 +1,367 @@
+            ? strtotime(str_replace('/', '-', $record['Finish date'] . ' ' . $record['Finish time']))
+            : null;
+
+            if ($session->datetimeknown && (empty($session->starttime) || empty($session->finishtime))) {
```
Multiple `if` conditions should be expanded to multi-line

NickLambellGCH replied:
Looks like this wasn't addressed

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -0,0 +1,367 @@
+
+            if ($session->datetimeknown && (empty($session->starttime) || empty($session->finishtime))) {
+                $this->errors[] = get_string('error:invaliddatetimedata', 'facetoface');
+                continue; // Skip if invalid.
```
Should be a new line before `continue`.

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -0,0 +1,367 @@
+            // Discount Cost: optional.
+            $session->discountcost = isset($record['Discount Cost'])
+                ? $record['Discount Cost']
+                : null;
```
Looks like these could be replaced with the `??` null coalesce operator.

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -0,0 +1,367 @@
+                $session->customfields = [
+                    $record['customfield_shortname'] => $record['customfield_value'],
+                ];
+            }
```
I don't think this custom field logic will work as it only allows a single custom field's value to be provided but there's also no handling to actually save the custom fields from what I can see since `$session` is just passed to `$DB->insert_record()`.

Typically, for custom fields in uploads we allow an arbitrary number of columns to be added with the 'customfield_<shortname>' heading format where the row values in that column are the value for that custom field. The code handling will be slightly different for F2F since it doesn't use Core Moodle custom fields, but it will likely follow similar logic.
I think the best way to handle this would be similar to [mod/signoff/import.php](https://github.com/gchlol/moodle-mod_signoff/blob/d661f8d0cfc863dc228b8d8225e27429dba2b2af/import.php#L286). You will effectively just get the full list of custom fields at the start of your import process, and then for every row, loop through the custom fields and see if the row has a matching 'customfield_<shortname>' column. If so, then store the value by key and save it after you create the session.

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -0,0 +1,367 @@
+                ];
+            }
+
+            $session->timecreated = time();
```
You should also set `timemodified`.

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -0,0 +1,367 @@
+
+            if (!$sessionid) {
+                $this->errors[] = get_string('error:failedtocreatesession', 'facetoface') . ' (' . implode(', ', $record) . ')';
+                continue;
```
Should have new line before `continue`.

NickLambellGCH replied:
Looks like this wasn't addressed

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -0,0 +1,367 @@
+            }
+
+            // Insert session date into mdl_facetoface_sessions_dates.
+            $sessionsdate = new \stdClass();
```
`stdClass` should be imported

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -0,0 +1,367 @@
+            $sessionsdate->timestart = $session->starttime;
+            $sessionsdate->timefinish = $session->finishtime;
+
+            if (!$DB->insert_record('facetoface_sessions_dates', $sessionsdate)) {
```
It's generally good to avoid having function calls in an `if` condition. This should store the function return to a variable like `$sessionid` does.

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -0,0 +1,367 @@
+                $this->errors[] = get_string('error:failedtoinsertdates', 'facetoface') . " for session ID {$sessionid}";
+            }
+        }
+        return empty($this->errors);
```
Should be a new line before `return`.

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -0,0 +1,367 @@
+                    $record['Start date and time'] = $date->format('Y-m-d H:i');
+                } else {
+                    $this->errors[] = [$index, get_string('error:invalidstarttime', 'facetoface')
+                    . ": {$record['Start date']} {$record['Start time']}"];
```
You should typically never concatenate language strings with other strings or values as the language strings may change or users / other languages may want the values in a different place.
Instead, you should use language string variables. In this case, your language string would become 'Finish date and time is invalid: {$a->date} {$a->time}', and you would pass an array or object to the `$a` param of `get_string()` with keys of 'date' and 'time', where the values are what get injected into the string.
There's a few cases of concatenation like this throughout this file that should be addressed 😄

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -0,0 +1,367 @@
+     *
+     * @return array array of error messages
+     */
+    public function get_errors() {
```
Function should have return type

NickLambellGCH replied:
Looks like this wasn't addressed

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -0,0 +1,367 @@
+     *
+     * @param int $fileid
+     *
+     * @return bool true if loaded successfully, false otherwise
```
PHPDoc descriptions should had a period at the end.

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -0,0 +1,367 @@
+     * Get parsed CSV records.
+     * @return array List of valid records.
+     */
+    public function get_records() {
```
Function should have return type

NickLambellGCH replied:
Looks like this wasn't addressed

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -0,0 +1,367 @@
+
+    /**
+     * Get parsed CSV records.
+     * @return array List of valid records.
```
Should be a new line before `@return`

NickLambellGCH commented on classes/form/bulk_session_confirm_form.php:
```diff
@@ -0,0 +1,63 @@
+use moodleform;
+
+defined('MOODLE_INTERNAL') || die();
+require_once($CFG->libdir.'/formslib.php');
```
Should be a new line before `require_once`

NickLambellGCH commented on classes/form/bulk_session_confirm_form.php:
```diff
@@ -0,0 +1,63 @@
+    /**
+     * Form definition.
+     */
+    public function definition() {
```
Function should have return type. `void` in this case

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -0,0 +1,367 @@
+            $sessionsdate->timefinish = $session->finishtime;
+
+            if (!$DB->insert_record('facetoface_sessions_dates', $sessionsdate)) {
+                $this->errors[] = get_string('error:failedtoinsertdates', 'facetoface') . " for session ID {$sessionid}";
```
Looks like this language string doesn't exist

NickLambellGCH commented on classes/form/bulk_session_confirm_form.php:
```diff
@@ -0,0 +1,63 @@
+    public function definition() {
+        $mform = $this->_form;
+
+        $f = $this->_customdata['f'] ?? 0;
```
Variable names should be descriptive as to what data they hold. Short variable names, especially single letter ones are usually always bad since they provide not context at all.
I know in this case it's because it comes from the URL param 'f' upstream but even in that case you can do `$facetofaceid = optional_param('f', ...`. Even more optimally, just name the param 'facetoface' since there's no reason it needs to be short, even though F2F itself seems to have a tendency to do that 😆 
I won't mention each case of the 'f' var as this comment should be used to correct all of them.

NickLambellGCH commented on classes/form/bulk_session_confirm_form.php:
```diff
@@ -0,0 +1,63 @@
+        $fileid = $this->_customdata['fileid'] ?? 0;
+
+        $mform->addElement('advcheckbox', 'suppressemail', get_string('suppressemail', 'facetoface'));
+        $mform->setType('suppressemail', PARAM_BOOL);
```
You shouldn't need `setType()` for checkboxes

Sajonasj-OC replied:
removed suppress email and case insensitive as it came from adding users to sessions. There are no users to be notified and emails to check for case insenstive.

NickLambellGCH commented on classes/form/bulk_session_confirm_form.php:
```diff
@@ -0,0 +1,63 @@
+            get_string('facetoface:confirmandprocess', 'mod_facetoface'));
+        $buttonarray[] = $mform->createElement('cancel');
+
+        $mform->addGroup($buttonarray, 'buttonar', '', ' ', false);
```
Since you just have a submit and cancel button, you should just be able to use `$this->add_action_buttons()` instead of building the button group yourself. It accepts a label for the submit button as the second param.

NickLambellGCH commented on classes/form/bulk_session_upload_form.php:
```diff
@@ -0,0 +1,107 @@
+namespace mod_facetoface\form;
+
+defined('MOODLE_INTERNAL') || die();
+require_once($CFG->dirroot.'/repository/lib.php');
```
Should be a new line above `require_once`.

NickLambellGCH commented on classes/form/bulk_session_upload_form.php:
```diff
@@ -0,0 +1,107 @@
+ * @copyright  GCHLOL, 2025
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class bulk_session_upload_form extends \moodleform {
```
Should import `moodleform`

NickLambellGCH replied:
This was imported but still has the `\`

NickLambellGCH commented on classes/form/bulk_session_upload_form.php:
```diff
@@ -0,0 +1,107 @@
+     * {@inheritDoc}
+     *
+     * @see \moodleform::definition()
+     */
```
You might see `@inheritDoc` in some places around the code (probably mostly from me 😆) but Moodle suggest against using it. Especially so in this case since the base `definition()` function doesn't have any useful information in the PHPDoc.
I think the standard "Form definition." doc like you have in the confirm form is probably good 😄

NickLambellGCH commented on classes/form/bulk_session_upload_form.php:
```diff
@@ -0,0 +1,107 @@
+        $mform->addElement('header', 'settingsheader', get_string('facetoface:uploadbulksessions', 'mod_facetoface'));
+
+        // Example CSV link.
+        $url = new \moodle_url('/mod/facetoface/example_bulk.csv');
```
`moodle_url` should be imported

NickLambellGCH commented on classes/form/bulk_session_upload_form.php:
```diff
@@ -0,0 +1,107 @@
+
+        // Example CSV link.
+        $url = new \moodle_url('/mod/facetoface/example_bulk.csv');
+        $link = \html_writer::link($url, 'example_bulk.csv');
```
`html_writer` should be imported

NickLambellGCH commented on classes/form/bulk_session_upload_form.php:
```diff
@@ -0,0 +1,107 @@
+        $mform->addElement('static', 'example_bulkcsv', get_string('facetoface:examplecsv', 'mod_facetoface'), $link);
+
+        // File manager for CSV upload.
+        $maxbytes = get_max_upload_file_size($CFG->maxbytes, 0);
```
Looks like `0` is already the default value for param 2.

NickLambellGCH commented on classes/form/bulk_session_upload_form.php:
```diff
@@ -0,0 +1,107 @@
+        $desc = get_string('facetoface:uploadsessionfiledesc', 'mod_facetoface');
+
+        // Split into rows by newlines.
+        $rows = explode("\n", $desc);
```
I like the ingenuity here, but I think having a language string be so dependent on a very specific format might be asking for trouble, especially since the field names are static in the code as well.
Instead of parsing out a language string I think you could just get away with a whole static table HTML or if you still wanted to construct it more dynamically, you could have a look at `html_writer::table()`.
Regardless of which method you choose I think the content could do with some refining as well since I find content like "Required: Format: HH:MM" a little bit confusing with the multiple '<label>:' parts. Instead, I think you should have a 'Required' column that is a simple yes or no and then a 'Format' column that can outline the expected values. e.g. 'HH:MM', 'yes / no', 'numeric value', etc.

NickLambellGCH commented on classes/form/bulk_session_upload_form.php:
```diff
@@ -0,0 +1,107 @@
+
+        // Additional checkbox for case-insensitive matching.
+        $mform->addElement('advcheckbox', 'caseinsensitive', get_string('caseinsensitive', 'mod_facetoface'));
+        $mform->setDefault('caseinsensitive', true);
```
`true`/`false` probably works here due to PHP's type conversions but `setDefault()` calls for checkboxes should always be either `0` or `1`.

NickLambellGCH commented on classes/form/bulk_session_upload_form.php:
```diff
@@ -0,0 +1,107 @@
+        $mform->addElement('hidden', 'validate', 1);
+        $mform->setType('validate', PARAM_INT);
+
+        $mform->addElement('submit', 'submit', get_string('facetoface:uploadandpreviewbulk', 'mod_facetoface'));
```
No cancel button? 😄

NickLambellGCH commented on classes/form/confirm_bookings_form.php:
```diff
@@ -21,7 +21,7 @@
 use single_button;
 
-defined('MOODLE_INTERNAL') || die;
+defined('MOODLE_INTERNAL') || die();
```
I think we should avoid changing existing F2F code unless it's causing issues

NickLambellGCH commented on db/access.php:
```diff
@@ -217,4 +217,15 @@
         )
     ),
+
+    // GCHLOL MF Ability to upload and process bookings via CSV.
```
The 'MF' part of this comment is the developer's initials so should be JS for you.
Comment should probably also mention that it's for sessions rather than bookings.

NickLambellGCH commented on db/access.php:
```diff
@@ -217,4 +217,15 @@
+            'editingteacher' => CAP_ALLOW,
+            'manager' => CAP_ALLOW,
+        ),
+    ),
```
`array(...)` should be replaced with the short form `[...]`

NickLambellGCH commented on example_bulk.csv:
Based off the code I don't think these headings are correct.
They will need to be adjusted based on the tweaks for case consistency and custom field handling that I suggested regardless 😄

NickLambellGCH commented on uploadbulksessions.php:
```diff
@@ -0,0 +1,173 @@
+
+if (!$facetoface = $DB->get_record('facetoface', ['id' => $f])) {
+        throw new moodle_exception('error:incorrectfacetofaceid', 'facetoface');
+}
```
If the F2F instance is expected to exist and will always have a lookup, then the 'f' URL param should probably be required instead of optional.

NickLambellGCH commented on uploadbulksessions.php:
```diff
@@ -0,0 +1,173 @@
+    $data = $mform->get_data();
+    $fileid = $data->csvfile ?: 0;
+
+    $mform = new bulk_session_confirm_form(null, ['f' => $f, 'fileid' => $fileid, 'caseinsensitive' => $caseinsensitive]);
```
Instead of redefining `$mform` multiple times, you should use different variable names. Since you always have the data for `bulk_session_upload_form` you can have something like `$uploadform = new bulk_session_upload_form(null, ['f' => $f])` before `if ($validate) {`  which you can use to get `$fileid` inside that `if` and for the display later in the page.

NickLambellGCH commented on uploadbulksessions.php:
```diff
@@ -0,0 +1,173 @@
+
+
+if ($validate) {
+    $heading = get_string('facetoface:validatebulksessions', 'facetoface');
```
Looks like this always gets overwritten

NickLambellGCH commented on uploadbulksessions.php:
```diff
@@ -0,0 +1,173 @@
+         $table->head = [get_string('uucsvline', 'tool_uploaduser'), get_string('status', 'facetoface')];
+
+        foreach ($errors as $error) {
+            if (is_array($error) && count($error) >= 2) {
+                $line = $error[0];
+                $messages = array_slice($error, 1);
+                foreach ($messages as $message) {
+                    $table->data[] = [$line, $message];
+                }
+            } else {
+                $table->data[] = ["-", is_string($error) ? $error : json_encode($error)];
+            }
+        }
+
+         echo html_writer::tag('div', html_writer::table($table), ['class' => 'flexible-wrap mb-4']);
+         echo $OUTPUT->footer();
+         exit;
+    } else {
+         echo $OUTPUT->header();
+         echo $OUTPUT->heading(get_string('facetoface:confirmbulkpreview', 'facetoface'), 3);
+
+         $records = $manager->get_records();
+
+        if (!empty($records)) {
+             $table = new html_table();
+             $table->attributes['class'] = 'f2fconfirmuploadlist m-auto generaltable mb-2';
+             $table->head = bulk_session_manager::get_headers();
+
+            foreach ($records as $record) {
+                 $table->data[] = array_values($record);
+            }
+             echo html_writer::tag('div', html_writer::table($table), ['class' => 'flexible-wrap mb-4']);
+        } else {
+             echo $OUTPUT->notification(get_string('facetoface:norecordsfound', 'facetoface'), 'info');
+        }
+
+         $mform->display();
+         echo $OUTPUT->footer();
+         exit;
+    }
```
A bit of the indentation in both halves of this `if` is too far in by 1 space.

NickLambellGCH commented on uploadbulksessions.php:
```diff
@@ -0,0 +1,173 @@
+    $errors = $manager->validate();
+
+    if (!empty($errors)) {
+         echo $OUTPUT->header();
```
This line is common to both halves of the `if` so should be extracted

NickLambellGCH commented on uploadbulksessions.php:
```diff
@@ -0,0 +1,173 @@
+         $table->head = [get_string('uucsvline', 'tool_uploaduser'), get_string('status', 'facetoface')];
+
+        foreach ($errors as $error) {
+            if (is_array($error) && count($error) >= 2) {
+                $line = $error[0];
+                $messages = array_slice($error, 1);
+                foreach ($messages as $message) {
+                    $table->data[] = [$line, $message];
+                }
+            } else {
+                $table->data[] = ["-", is_string($error) ? $error : json_encode($error)];
+            }
```
Instead of using `if`/`else` for flow control in `foreach` loops you should use `if`/`continue` to reduce indentation. In this case, since the `else` is the simple portion, I'd flip the `if` and use `continue` as suggested. e.g.
```php
if (
    !is_array($error) ||
    count($error) < 2
) {
    $table->data[] = ["-" ...

    continue;
}

$line = $error[0];
...
```

NickLambellGCH commented on uploadbulksessions.php:
```diff
@@ -0,0 +1,173 @@
+
+         echo html_writer::tag('div', html_writer::table($table), ['class' => 'flexible-wrap mb-4']);
+         echo $OUTPUT->footer();
+         exit;
```
Should have new line before `exit`

NickLambellGCH commented on uploadbulksessions.php:
```diff
@@ -0,0 +1,173 @@
+         echo html_writer::tag('div', html_writer::table($table), ['class' => 'flexible-wrap mb-4']);
+         echo $OUTPUT->footer();
+         exit;
+    } else {
```
Should have new line before `else`

NickLambellGCH commented on uploadbulksessions.php:
```diff
@@ -0,0 +1,173 @@
+
+            foreach ($records as $record) {
+                 $table->data[] = array_values($record);
+            }
```
New line after closing of the `foreach` block / before `echo`

NickLambellGCH commented on uploadbulksessions.php:
```diff
@@ -0,0 +1,173 @@
+                 $table->data[] = array_values($record);
+            }
+             echo html_writer::tag('div', html_writer::table($table), ['class' => 'flexible-wrap mb-4']);
+        } else {
```
New line before `else`

NickLambellGCH commented on uploadbulksessions.php:
```diff
@@ -0,0 +1,173 @@
+
+         $mform->display();
+         echo $OUTPUT->footer();
+         exit;
```
New line before `exit`

NickLambellGCH commented on uploadbulksessions.php:
```diff
@@ -0,0 +1,173 @@
+    }
+}
+
+if ($process && $fileid && $f) {
```
Each `if` condition should be on its own line

NickLambellGCH commented on uploadbulksessions.php:
```diff
@@ -0,0 +1,173 @@
+        $manager->suppress_email();
+    }
+
+     $errors = $manager->validate();
+
+    if (empty($errors)) {
+         $manager->process();
+
+         $params = [
+             'context'  => $modulecontext,
+             'objectid' => $f,
+         ];
+         $event = \mod_facetoface\event\csv_processed::create($params);
+         $event->add_record_snapshot('facetoface', $facetoface);
+         $event->trigger();
+
+         redirect(
+             new moodle_url('/mod/facetoface/uploadbulksessions.php', ['f' => $f]),
+             get_string('facetoface:bulksessionsprocessed', 'mod_facetoface'),
+             null,
+             notification::NOTIFY_SUCCESS
+         );
+    }
+
+     // If errors exist, redirect back with error message.
+     $errmsg = get_string('error:bulkuploadfileerrorsfound', 'mod_facetoface', count($errors));
+     redirect(
+         new moodle_url('/mod/facetoface/uploadbulksessions.php', ['f' => $f]),
+         $errmsg,
+         null,
+         notification::NOTIFY_ERROR
+     );
+}
```
Indentation is generally off here too

NickLambellGCH commented on uploadbulksessions.php:
```diff
@@ -0,0 +1,173 @@
+    if (empty($errors)) {
+         $manager->process();
+
+         $params = [
+             'context'  => $modulecontext,
+             'objectid' => $f,
+         ];
+         $event = \mod_facetoface\event\csv_processed::create($params);
+         $event->add_record_snapshot('facetoface', $facetoface);
+         $event->trigger();
```
You should be careful when reusing events that were potentially designed for something else. In this case even though the event is fairly generically named, if you look at the description, if specifically mentions uploading session bookings.
We should probably add our own even for this upload instead as it's useful to have logged

Sajonasj-OC replied:
noted. will create a new event file.

Sajonasj-OC replied:
![image](https://github.com/user-attachments/assets/fda60674-69f7-4e1c-b23c-54181e9d0ea4)

Do i need to change anyting else on the event log?

NickLambellGCH replied:
You could potentially change the event name to make it more distinct otherwise looks good to me 😄

Sajonasj-OC replied:
![image](https://github.com/user-attachments/assets/062c4a75-fc25-4ac7-aaca-50fc3d461592)

NickLambellGCH commented on uploadbulksessions.php:
```diff
@@ -0,0 +1,173 @@
+$heading = get_string('facetoface:uploadbulksessions', 'facetoface');
+$mform = new bulk_session_upload_form(null, ['f' => $f]);
+$mform->set_data(['f' => $f, 'validate' => 1]);
+$PAGE->set_pagelayout('standard');
+$PAGE->set_title($heading);
+$PAGE->set_heading($heading);
```
I'll usually place all the `$PAGE` metadata setting before any page logic where possible.

NickLambellGCH commented on uploadbulksessions.php:
```diff
@@ -0,0 +1,173 @@
+$PAGE->set_heading($heading);
+
+echo $OUTPUT->header();
+$mform->display();
```
New line before and after this line is good to more distinctly separate the header and footer from body content.

NickLambellGCH commented on view.php:
```diff
@@ -240,6 +241,7 @@ function print_session_list($courseid, $facetoface, $location) {
             $viewattendees,
             $editsessions,
             !$bulksignup,
+            $uploadbulksessions,
```
Adding this here doesn't seem correct as `print_session_list_table()` hasn't been extended to take extra params

NickLambellGCH commented on view.php:
```diff
@@ -253,6 +255,14 @@ function print_session_list($courseid, $facetoface, $location) {
     }
 
+    if ($uploadbulksessions) {
+        $adduploadlink = html_writer::link(
```
This variable should be distinctly named so that the later `$adduploadlink` doesn't reassign it

NickLambellGCH reviewed (changes requested):
I've done my initial pass over this one but haven't tested it out fully yet. Don't be too thrown off by the amount of feedback either. I'm typically more picky early on to help with learning code style, syntax, best practices, etc. 😄 
One thing to be careful of is to not just copy existing code verbatim as often times it can be really old and most likely not follow good coding standards or just be poorly written. 😢 

Since you're adding a capability in `access.php` you'll have to do a version number bump. It should just be a bump to the increment part of the number since it's an external plugin.

Sajonasj-OC commented:
Possibly need more validation rules. It does not filter duplicate sessions as every sessions added generate new session id's, need to explore the rules regarding duplicates.

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -0,0 +1,469 @@
+    private bool $usefile = false;
+
+    /** @var stored_file|null Reference to uploaded file */
+    private ?\stored_file $file = null;
```
`stored_file` should be imported

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -0,0 +1,469 @@
+     */
+    private function get_iterator(): Generator {
+        if (!$this->usefile) {
+
```
You shouldn't need a blank line at the start of control structures like this. Only when keyword blocks e.g. `foreach` directly follow other content within the same control structure / indentation.

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -0,0 +1,469 @@
+        $headerline = fgetcsv($handle, $maxlinelength, $delimiter);
+
+        if (
+            $headerline === false ||
```
This is technically a bit of a duplicate check since `empty()` will also check for `=== false`.

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -0,0 +1,469 @@
+            ) {
+            fclose($handle);
+            throw new moodle_exception('error:bookingsuploadfileheaderfieldmismatch', 'mod_facetoface',
+                get_string('error:noheaderrow', 'facetoface'));
```
For wrapping long statements like this, it's a lot more readable to just place each parameter on its own line separate from the opening and closing parenthesis too

NickLambellGCH replied:
Looks like this string doesn't exist. I usually find it helpful to add the string to the language file right when I reference it so that I don't miss anything 😄

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -0,0 +1,469 @@
+            empty($headerline)
+            ) {
+            fclose($handle);
+            throw new moodle_exception('error:bookingsuploadfileheaderfieldmismatch', 'mod_facetoface',
```
new line before `throw` would be nice since it's technically a flow control keyword

NickLambellGCH replied:
I think reusing 'error:bookingsuploadfileheaderfieldmismatch' here isn't really the best move since it specifically mentions a "Mismatched number of fields" but that is not the case.
You're also running into a case of effectively concatenating language strings by passing a language string into the `$a` parameter. Although in this case it's actually being passed to the `$link` param which I suspect wasn't the intention.

You should instead just implement the 'error:noheaderrow' string and use that directly as the exception's `$errorcode`.

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -0,0 +1,469 @@
+                    new moodle_url('/mod/facetoface/view.php', ['f' => $this->facetofaceid]),
+                    "Missing required column: {$required}"
+                );
+            }
```
`if` statements that span the entire content of their outer block are good to avoid. Since this is a `foreach`, you should just convert this check to a `if(in_array ...` with a `continue`.

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -0,0 +1,469 @@
+                throw new moodle_exception(
+                    'error:bookingsuploadfileheaderfieldmismatch',
+                    'mod_facetoface',
+                    new moodle_url('/mod/facetoface/view.php', ['f' => $this->facetofaceid]),
```
Should this instead link back to `uploadbulksessions.php`?

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -0,0 +1,469 @@
+            if (!in_array($required, $headerline, true)) {
+                fclose($handle);
+                throw new moodle_exception(
+                    'error:bookingsuploadfileheaderfieldmismatch',
```
This should be updated to its own language string as above. Something that specifically mentions missing required header X

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -0,0 +1,469 @@
+        foreach ($headerline as &$field) {
+            $field = trim($field, "'");
+        }
+        unset($field);
```
Array value modification by reference like this is definitely something to avoid since it's not really intuitive when skimming. Instead, for the best readability you should store the raw line to an array like `$rawheaderline` and then write it to the final `$headerline` array in the loop. Or even just use something like `array_map()` for simple cases like this.

Looking at this however does make me wonder why the `trim()` filtering is required in the first place. I think the main difference vs other importers that we've written is `fgetcsv()` vs `str_getcsv()` but the handling is a bit different.

Sajonasj-OC replied:
The trim function was originally added because the CSV file required single quotes around its date fields. This prevented those dates from breaking when the file was opened in certain spreadsheet programs. Removed it cause I split the dates.

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -0,0 +1,469 @@
+
+        try {
+            while (($data = fgetcsv($handle, $maxlinelength, $delimiter)) !== false) {
+
```
Shouldn't need blank line here

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -0,0 +1,469 @@
+            while (($data = fgetcsv($handle, $maxlinelength, $delimiter)) !== false) {
+
+                if (count($data) !== count($headerline)) {
+                    throw new moodle_exception('error:invalidcsvrow', 'facetoface');
```
This language string doesn't exist, but this is probably the perfect spot to actually use 'error:bookingsuploadfileheaderfieldmismatch'. You'll just need an incrementing value to count the row number 😄

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -0,0 +1,469 @@
+                foreach ($data as &$field) {
+                    $field = trim($field, "'");
+                }
+                unset($field);
```
Pending potential discussion on the above similar usage.

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -0,0 +1,469 @@
+                unset($field);
+
+                $row = array_combine($headerline, $data);
+                yield $row;
```
I don't think this needs to be split out to a variable that is immediately returned for a simple function call like this one

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -0,0 +1,469 @@
+
+                $row = array_combine($headerline, $data);
+                yield $row;
+
```
Shouldn't need a new line here

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -0,0 +1,469 @@
+                yield $row;
+
+            }
+
```
Shouldn't need a new line here since it's just a block closure on the line above the `finally` keyword

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -0,0 +1,469 @@
+        }
+
+        foreach ($this->records as $index => $record) {
+
```
Shouldn't need blank line

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -0,0 +1,469 @@
+            if (!$startdt) {
+                $params = (object)[
+                    'date' => $record['Start Date'],
+                    'time' => $record['Start Time']
```
Multi-line arrays should have a trailing comma

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -0,0 +1,469 @@
+
+                $this->errors[] = [
+                    $index,
+                    get_string('error:invalidstarttime', 'facetoface', $params)
```
Should have trailing comma

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -0,0 +1,469 @@
+            }
+
+            $finishdt = DateTime::createFromFormat('d/m/Y H:i', $record['Finish Date'].' '.$record['Finish Time']);
+            $finishdt = DateTime::createFromFormat('d/m/Y H:i', $record['Finish Date'].' '.$record['Finish Time']);
```
This line is duplicated

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -0,0 +1,469 @@
+            if (!$finishdt) {
+                $params = (object)[
+                    'date' => $record['Finish Date'],
+                    'time' => $record['Finish Time']
```
Should have a trailing comma

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -0,0 +1,469 @@
+
+                $this->errors[] = [
+                    $index,
+                    get_string('error:invalidfinishtime', 'facetoface', $params)
```
Should have a trailing comma

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -0,0 +1,469 @@
+            // Allow Cancellations (required, "yes" or "no").
+            $allowcancel = strtolower($record['Allow Cancellations'] ?? '');
+            if (!in_array($allowcancel, ['yes', 'no'], true)) {
+                $this->errors[] = [$index, get_string('error:invalidallowcancel', 'facetoface')];
```
Looks like this string doesn't exist

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -0,0 +1,469 @@
+     *
+     * @return bool True if all sessions were created successfully, false otherwise.
+     */
+    public function process() {
```
Function should have a return type

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -0,0 +1,469 @@
+            $session->facetoface = $this->facetofaceid;
+
+            $session->datetimeknown = 1;
+            if (!empty($record['Session Date/Time Known']) &&
```
The first condition should be on its own line

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -0,0 +1,469 @@
+            }
+
+            $session->starttime = null;
+            if (!empty($record['Start Date']) && !empty($record['Start Time'])) {
```
Should be expanded to multi-line

Sajonasj-OC replied:
I fixed this section last time I remember. It must be a formatter that moodle suggested.

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -0,0 +1,469 @@
+            }
+
+            $session->finishtime = null;
+            if (!empty($record['Finish Date']) && !empty($record['Finish Time'])) {
```
Should be expanded to multi-line

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -0,0 +1,469 @@
+            }
+
+            $session->allowcancel = 1;
+            if (!empty($record['Allow Cancellations']) && $record['Allow Cancellations'] === 'no') {
```
Should be expanded to multi-line

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -0,0 +1,469 @@
+            }
+
+            $session->capacity = 10;
+            if (!empty($record['Capacity']) && is_numeric($record['Capacity'])) {
```
Should be expanded to multi-line

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -0,0 +1,469 @@
+            }
+
+            $session->overbook = 1;
+            if (!empty($record['Allow Overbookings']) && $record['Allow Overbookings'] === 'no') {
```
Should be expanded to multi-line

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -0,0 +1,469 @@
+            }
+
+            $session->duration = 0;
+            if (!empty($record['Duration']) && is_numeric($record['Duration'])) {
```
Should be expanded to multi-line

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -0,0 +1,469 @@
+            }
+
+            $session->normalcost = null;
+            if (!empty($record['Normal Cost']) && is_numeric($record['Normal Cost'])) {
```
Should be expanded to multi-line

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -0,0 +1,469 @@
+            }
+
+            $session->discountcost = null;
+            if (!empty($record['Discount Cost']) && is_numeric($record['Discount Cost'])) {
```
Should be expanded to multi-line

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -0,0 +1,469 @@
+            }
+
+            $session->details = '';
+            if (!empty($record['Details']) && is_string($record['Details'])) {
```
Should be expanded to multi-line

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -0,0 +1,469 @@
+            $sessionsdateid = $DB->insert_record('facetoface_sessions_dates', $sessionsdate);
+
+            if (!$sessionsdateid) {
+                $this->errors[] = get_string('error:failedtoinsertdates', 'facetoface') . " (ID: $sessionid)";
```
Shouldn't concatenate with language strings, instead use `$a` syntax. I think the content that's currently concatenated should also be updated to mention that the ID is the session ID. e.g. 'Failed to create dates for session #$a'.
Also note the 'create' vs 'insert' language since these are user facing errors 😄

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -0,0 +1,469 @@
+                            $this->errors[] = get_string('error:couldnotsavecustomfield', 'facetoface')." ($shortname)";
+                        }
+                    }
+                }
```
Instead of nested `if`s, you should use `continue` to control the flow

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -0,0 +1,469 @@
+                    if (isset($customfieldsbyshortname[$shortname])) {
+                        $field = $customfieldsbyshortname[$shortname];
+                        if (!facetoface_save_customfield_value($field->id, $value, $sessionid, 'session')) {
+                            $this->errors[] = get_string('error:couldnotsavecustomfield', 'facetoface')." ($shortname)";
```
This is concatenating with a language string. Since we don't want to modify the existing string from the base F2F code, you should create your own with `$a` syntax.
I also just noticed that values added to `$this->errors` should themselves be arrays containing the line index and the error message but the errors in `process()` don't seem to do that.

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -0,0 +1,469 @@
+        if ($this->usefile) {
+            $this->records = iterator_to_array($this->get_iterator());
+        }
+        return $this->records;
```
Should be a new line before `return`.

NickLambellGCH commented on classes/event/csv_processed_bulksession.php:
```diff
@@ -0,0 +1,81 @@
+ * @author      Jonas Sajonas
+ * @license     https://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class csv_processed_bulksession extends \core\event\base {
```
`base` should be imported

NickLambellGCH commented on classes/event/csv_processed_bulksession.php:
```diff
@@ -0,0 +1,81 @@
+     *
+     * @return void
+     */
+    protected function init() {
```
Function should have return type

NickLambellGCH commented on classes/event/csv_processed_bulksession.php:
```diff
@@ -0,0 +1,81 @@
+     *
+     * @return string
+     */
+    public function get_description() {
```
Function should have return type

NickLambellGCH commented on classes/event/csv_processed_bulksession.php:
```diff
@@ -0,0 +1,81 @@
+     *
+     * @return string
+     */
+    public static function get_name() {
```
Function should have return type

NickLambellGCH commented on classes/event/csv_processed_bulksession.php:
```diff
@@ -0,0 +1,81 @@
+     *
+     * @return \moodle_url
+     */
+    public function get_url() {
```
Function should have return type

NickLambellGCH commented on classes/event/csv_processed_bulksession.php:
```diff
@@ -0,0 +1,81 @@
+     * @throws \coding_exception
+     * @return void
+     */
+    protected function validate_data() {
```
Function should have return type

NickLambellGCH commented on classes/event/csv_processed_bulksession.php:
```diff
@@ -0,0 +1,81 @@
+     * @return \moodle_url
+     */
+    public function get_url() {
+        return new \moodle_url('/mod/facetoface/uploadbulksessions.php', ['f' => $this->objectid]);
```
`moodle_url` should be imported

NickLambellGCH commented on classes/event/csv_processed_bulksession.php:
```diff
@@ -0,0 +1,81 @@
+        parent::validate_data();
+
+        if ($this->contextlevel != CONTEXT_MODULE) {
+            throw new \coding_exception('Context level must be CONTEXT_MODULE.');
```
`coding_exception` should be imported

NickLambellGCH commented on classes/form/bulk_session_confirm_form.php:
```diff
@@ -0,0 +1,65 @@
+        $mform = $this->_form;
+
+        // Retrieve the new parameter f2did from custom data.
+        $f2did = $this->_customdata['f'] ?? 0;
```
Should this be `$f2fid`?
Would also be good to see the `_customdata` key updated to `f2fid` for consistency. Related comments later on the code where it's set.

Sajonasj-OC replied:
Do you mean the f2did? or the 'f'? I changed the f2did to f2fid cause it's a typo.

NickLambellGCH replied:
I meant the typo but I'm guessing the 'f' will be changing too 😄

NickLambellGCH commented on classes/form/bulk_session_confirm_form.php:
```diff
@@ -0,0 +1,65 @@
+
+        // Add hidden fields using the new name.
+        $mform->addElement('hidden', 'f', $f2did);
+        $mform->setType('f', PARAM_INT);
```
This will change to 'f2fid' as well

NickLambellGCH commented on classes/form/bulk_session_upload_form.php:
```diff
@@ -0,0 +1,178 @@
+            get_string('upload_format', 'mod_facetoface')
+        ];
+
+        // 1) Session Date/Time Known
```
Looks like this is the only row that's labelled with a comment. I don't mind either way, but they should probably all be labelled or none labelled.
For maintainability, if you do keep the comments, you could probably not have the number portion 😄

NickLambellGCH commented on classes/form/bulk_session_upload_form.php:
```diff
@@ -0,0 +1,178 @@
+        $table->head = [
+            get_string('upload_field', 'mod_facetoface'),
+            get_string('upload_required', 'mod_facetoface'),
+            get_string('upload_format', 'mod_facetoface')
```
Multi-line arrays should have a trailing comma.

NickLambellGCH commented on classes/form/bulk_session_upload_form.php:
```diff
@@ -0,0 +1,178 @@
+        $table->data[] = [
+            get_string('upload_field_sessiondatetime', 'mod_facetoface'),
+            get_string('upload_req_yesno', 'mod_facetoface'),
+            get_string('upload_format_na', 'mod_facetoface')
```
Multi-line arrays should have a trailing comma.

NickLambellGCH commented on classes/form/bulk_session_upload_form.php:
```diff
@@ -0,0 +1,178 @@
+        $table->data[] = [
+            get_string('upload_field_startdate', 'mod_facetoface'),
+            get_string('upload_req_yes', 'mod_facetoface'),
+            get_string('upload_format_date', 'mod_facetoface')
```
Multi-line arrays should have a trailing comma.

NickLambellGCH commented on classes/form/bulk_session_upload_form.php:
```diff
@@ -0,0 +1,178 @@
+        $table->data[] = [
+            get_string('upload_field_starttime', 'mod_facetoface'),
+            get_string('upload_req_yes', 'mod_facetoface'),
+            get_string('upload_format_time', 'mod_facetoface')
```
Multi-line arrays should have a trailing comma.

NickLambellGCH commented on classes/form/bulk_session_upload_form.php:
```diff
@@ -0,0 +1,178 @@
+        $table->data[] = [
+            get_string('upload_field_finishdate', 'mod_facetoface'),
+            get_string('upload_req_yes', 'mod_facetoface'),
+            get_string('upload_format_date', 'mod_facetoface')
```
Multi-line arrays should have a trailing comma.

NickLambellGCH commented on classes/form/bulk_session_upload_form.php:
```diff
@@ -0,0 +1,178 @@
+        $table->data[] = [
+            get_string('upload_field_finishtime', 'mod_facetoface'),
+            get_string('upload_req_yes', 'mod_facetoface'),
+            get_string('upload_format_time', 'mod_facetoface')
```
Multi-line arrays should have a trailing comma.

NickLambellGCH commented on classes/form/bulk_session_upload_form.php:
```diff
@@ -0,0 +1,178 @@
+        $table->data[] = [
+            get_string('upload_field_allowcancellations', 'mod_facetoface'),
+            get_string('upload_req_yesno', 'mod_facetoface'),
+            get_string('upload_format_na', 'mod_facetoface')
```
Multi-line arrays should have a trailing comma.

NickLambellGCH commented on classes/form/bulk_session_upload_form.php:
```diff
@@ -0,0 +1,178 @@
+        $table->data[] = [
+            get_string('upload_field_capacity', 'mod_facetoface'),
+            get_string('upload_req_yes', 'mod_facetoface'),
+            get_string('upload_format_numeric', 'mod_facetoface')
```
Multi-line arrays should have a trailing comma.

NickLambellGCH commented on classes/form/bulk_session_upload_form.php:
```diff
@@ -0,0 +1,178 @@
+        $table->data[] = [
+            get_string('upload_field_allowoverbookings', 'mod_facetoface'),
+            get_string('upload_req_yesno', 'mod_facetoface'),
+            get_string('upload_format_na', 'mod_facetoface')
```
Multi-line arrays should have a trailing comma.

NickLambellGCH commented on classes/form/bulk_session_upload_form.php:
```diff
@@ -0,0 +1,178 @@
+        $table->data[] = [
+            get_string('upload_field_duration', 'mod_facetoface'),
+            get_string('upload_req_yes', 'mod_facetoface'),
+            get_string('upload_format_numericmins', 'mod_facetoface')
```
Multi-line arrays should have a trailing comma.

NickLambellGCH commented on classes/form/bulk_session_upload_form.php:
```diff
@@ -0,0 +1,178 @@
+        $table->data[] = [
+            get_string('upload_field_normalcost', 'mod_facetoface'),
+            get_string('upload_req_optional', 'mod_facetoface'),
+            get_string('upload_format_numeric', 'mod_facetoface')
```
Multi-line arrays should have a trailing comma.

NickLambellGCH commented on classes/form/bulk_session_upload_form.php:
```diff
@@ -0,0 +1,178 @@
+        $table->data[] = [
+            get_string('upload_field_discountcost', 'mod_facetoface'),
+            get_string('upload_req_optional', 'mod_facetoface'),
+            get_string('upload_format_numeric', 'mod_facetoface')
```
Multi-line arrays should have a trailing comma.

NickLambellGCH commented on classes/form/bulk_session_upload_form.php:
```diff
@@ -0,0 +1,178 @@
+        $table->data[] = [
+            get_string('upload_field_details', 'mod_facetoface'),
+            get_string('upload_req_optional', 'mod_facetoface'),
+            get_string('upload_format_string', 'mod_facetoface')
```
Multi-line arrays should have a trailing comma.

NickLambellGCH commented on classes/form/bulk_session_upload_form.php:
```diff
@@ -0,0 +1,178 @@
+        $table->data[] = [
+            get_string('upload_field_customfield', 'mod_facetoface'),
+            get_string('upload_req_optional', 'mod_facetoface'),
+            get_string('upload_format_customfield', 'mod_facetoface')
```
Multi-line arrays should have a trailing comma.

NickLambellGCH commented on classes/form/bulk_session_upload_form.php:
```diff
@@ -0,0 +1,178 @@
+            get_string('upload_format_customfield', 'mod_facetoface')
+        ];
+
+        $tablehtml = html_writer::table($table);
```
For code separation and readability, I think it would make sense to move the generation of this table into a private function on this class that just returns the string.

NickLambellGCH commented on classes/form/bulk_session_upload_form.php:
```diff
@@ -0,0 +1,178 @@
+
+        $f2fid = $this->_customdata['f'] ?? 0;
+        $mform->addElement('hidden', 'f', $f2fid);
+        $mform->setType('f', PARAM_INT);
```
This will all change to 'f2fid' with my earlier suggested changes

NickLambellGCH commented on classes/form/upload_bookings_form.php:
I didn't notice that these changes were to an existing F2F file in my previous review. Since we want to avoid modifying the existing code where possible, I'll get you to revert your changes to this file.

Sajonasj-OC replied:
I think I've mistakenly changed this while editing. reverted back.

NickLambellGCH commented on db/access.php:
```diff
@@ -217,4 +217,16 @@
+    'archetypes' => [
+        'teacher' => CAP_ALLOW,
+        'editingteacher' => CAP_ALLOW,
+        'manager' => CAP_ALLOW,
```
It looks like in converting to the short array syntax, the indentation on these lines was broken.

NickLambellGCH commented on lang/en/facetoface.php:
```diff
@@ -857,11 +857,77 @@
+$string['upload_format_time'] = 'HH:MM';
+$string['upload_format_numeric'] = 'Numeric';
+$string['upload_format_numericmins'] = 'Numeric (minutes)';
+$string['upload_format_string'] = 'Any text';
+$string['upload_format_customfield'] = 'Any text';
```
Language strings typically shouldn't have underscores outside of a few exceptions like '_desc' for form field help text. The categorisation comments in the language file should also only really be used to categorise into the top-level categories like the existing ones.

Since the strings here are all related and eventually output to the 'csvuploadhelp' field it probably makes sense to prefix them all with that and then part of the keys you already have. e.g. 'upload_field_sessiondatetime' becomes 'csvuploadhelp:fieldsessiondatetime'.
Seeing as the other headers and the related values are pretty consistent, I'd also just expand the 'req' shortening to the full 'required'.

NickLambellGCH replied:
Since 'upload_format_string' and 'upload_format_customfield' have the same value and don't seem like they would ever need to be different, the place that uses the 'customfield' format should just use string.

NickLambellGCH commented on lang/en/facetoface.php:
```diff
@@ -857,11 +857,77 @@
+
+// Table headers.
+$string['upload_field'] = 'Field';
+$string['upload_required'] = 'Required?';
```
This probably doesn't need the question mark

NickLambellGCH commented on lang/en/facetoface.php:
```diff
@@ -857,11 +857,77 @@
+
+$string['examplecsv'] = 'example.csv';
+$string['error:invalidstarttime'] = 'Start date/time is invalid: {$a->date} {$a->time}';
+$string['error:invalidfinishtime'] = 'Finish date/time is invalid: {$a->date} {$a->time}';
```
These keys duplicate those in the block above and should probably replace them.

NickLambellGCH commented on lang/en/facetoface.php:
```diff
@@ -857,11 +857,77 @@
+$string['upload_req_yesno'] = 'No';
+$string['upload_req_yes'] = 'Yes';
+$string['upload_req_no'] = 'No';
+$string['upload_req_optional'] = 'Optional';
```
Isn't 'optional' just the same as not required? The required column should probably just be either 'yes' or 'no'. For which it looks like there are existing language strings from core. e.g. `get_string('yes')`.

NickLambellGCH commented on lang/en/facetoface.php:
```diff
@@ -857,11 +857,77 @@
+$string['upload_field_customfield'] = 'Customfield_<suffix>';
+
+// Required? values.
+$string['upload_req_yesno'] = 'No';
```
I think the wires might have gotten a bit crossed here. The fields that are using this for their required column also seem to have format as 'N/A' but their format should be 'yes / no' and the required column either the yes or the no option depending on if it is required or not.

NickLambellGCH commented on lang/en/facetoface.php:
```diff
@@ -857,11 +857,77 @@
+$string['upload_field_normalcost'] = 'Normal Cost';
+$string['upload_field_discountcost'] = 'Discount Cost';
+$string['upload_field_details'] = 'Details';
+$string['upload_field_customfield'] = 'Customfield_<suffix>';
```
_Technically_ the 'Customfield_' part is the prefix here and the 'suffix' part is the value. It also looks like the `<suffix>` part is being parsed out as HTML by the browser and so isn't visible.
I like the use of the `<>` to indicate a placeholder but in this case it doesn't really give context to the user of what the value should be. Typically, our admin users are fairly versed with what 'shortname' means so something like `Customfield_<shortname>` would probably work well. To have the language string not parsed as HTML you should be able to use `&lt;` and `&gt;`.

NickLambellGCH commented on uploadbulksessions.php:
```diff
@@ -0,0 +1,203 @@
+use mod_facetoface\form\bulk_session_confirm_form;
+use mod_facetoface\bulk_session_manager;
+
+$f = required_param('f', PARAM_INT);
```
As from related feedback, this should be `$f2fid`. You could keep the URL parameter itself as 'f' but something like 'facetoface' or matching 'f2fid' would be more typical 😄

NickLambellGCH commented on uploadbulksessions.php:
```diff
@@ -0,0 +1,203 @@
+$uploadform = new bulk_session_upload_form(null, ['f' => $f]);
+
+// Helper function to display errors and exit.
+function handle_bulk_upload_errors($errors) {
```
In file functions should still meet requirements e.g. PHPDoc, param types, return type.

NickLambellGCH commented on uploadbulksessions.php:
```diff
@@ -0,0 +1,203 @@
+    $table = new html_table();
+    $table->attributes['class'] = 'f2fbookingsuploadlist m-auto generaltable mb-2';
+    $table->head = [
+        get_string('uucsvline', 'tool_uploaduser'),
```
Unless explicitly defined as a dependency and only for cases where the relation makes sense, you should avoid using language strings from other plugins.

NickLambellGCH commented on uploadbulksessions.php:
```diff
@@ -0,0 +1,203 @@
+    $table->attributes['class'] = 'f2fbookingsuploadlist m-auto generaltable mb-2';
+    $table->head = [
+        get_string('uucsvline', 'tool_uploaduser'),
+        get_string('status', 'facetoface')
```
Multi-line arrays should have a trailing comma

NickLambellGCH commented on uploadbulksessions.php:
```diff
@@ -0,0 +1,203 @@
+    global $OUTPUT;
+
+    echo $OUTPUT->header();
+    echo $OUTPUT->notification(get_string('error:bulkuploadfileerrorsfound', 'mod_facetoface', count($errors)), 'error');
```
Calls to `$OUTPUT->notification()` should use the `core\output\notification::NOTIFY_x` (imported as `notification`) constants for the type, same as `redirect()` that you have later on

NickLambellGCH replied:
This language string references bookings not sessions, you should probably make a copy of it for your usage.

NickLambellGCH commented on uploadbulksessions.php:
```diff
@@ -0,0 +1,203 @@
+    ];
+
+    foreach ($errors as $error) {
+        if (!is_array($error) || count($error) < 2) {
```
Should be expanded to multi-line `if`

NickLambellGCH commented on uploadbulksessions.php:
```diff
@@ -0,0 +1,203 @@
+    foreach ($errors as $error) {
+        if (!is_array($error) || count($error) < 2) {
+            $table->data[] = ["-", is_string($error) ? $error : json_encode($error)];
+            continue;
```
Should have a new line before `continue`

NickLambellGCH commented on uploadbulksessions.php:
```diff
@@ -0,0 +1,203 @@
+
+    if ($uploadform->is_cancelled()) {
+        redirect(new moodle_url('/mod/facetoface/view.php', ['f' => $f]));
+        exit;
```
Should have new line before `exit`

NickLambellGCH commented on uploadbulksessions.php:
```diff
@@ -0,0 +1,203 @@
+        foreach ($records as $record) {
+            $table->data[] = array_values($record);
+        }
+        echo html_writer::tag('div', html_writer::table($table), ['class' => 'flexible-wrap mb-4']);
```
Should have a new line before `echo` because it is after a block

NickLambellGCH commented on uploadbulksessions.php:
```diff
@@ -0,0 +1,203 @@
+    // Display the confirm form.
+    $confirmform->display();
+    echo $OUTPUT->footer();
+    exit;
```
Should have a new line before `exit`

NickLambellGCH commented on uploadbulksessions.php:
```diff
@@ -0,0 +1,203 @@
+    $process &&
+    $fileid &&
+    $f
+    ) {
```
This line shouldn't be indented

NickLambellGCH commented on uploadbulksessions.php:
```diff
@@ -0,0 +1,203 @@
+    $errors = $manager->validate();
+
+    if (empty($errors)) {
+         $manager->process();
+
+         $params = [
+             'context'  => $modulecontext,
+             'objectid' => $f,
+         ];
+
+         $event = \mod_facetoface\event\csv_processed_bulksession::create($params);
+         $event->add_record_snapshot('facetoface', $facetoface);
+         $event->trigger();
+
+         redirect(
+             new moodle_url('/mod/facetoface/uploadbulksessions.php', ['f' => $f]),
+             get_string('facetoface:bulksessionsprocessed', 'mod_facetoface'),
+             null,
+             notification::NOTIFY_SUCCESS
+         );
```
Looks like this block is indented 1 space too many

NickLambellGCH commented on uploadbulksessions.php:
```diff
@@ -0,0 +1,203 @@
+             'objectid' => $f,
+         ];
+
+         $event = \mod_facetoface\event\csv_processed_bulksession::create($params);
```
`csv_processed_bulksession` should be imported

NickLambellGCH reviewed (changes requested):
Thanks for working through that previous set of feedback, definitely liking the more structural changes 😄 
This is a bit of another big batch but it's mostly syntax and semantics plus a couple of things that were missed in the last batch. Probably buried in the big list 😆

Sajonasj-OC commented:
Thanks for reviewing this, Nick! This PR is the first one containing most of the data manipulation, so it was a bit dizzying to work on. I’ve resolved the issues you highlighted and taken another pass to see if I missed anything.

I also changed the table structure and fixed a bug involving missing custom field headers. I might still adjust those headers later, depending on whether we decide to keep them permanently on the table or make them optional.

Next, I’ll pull these changes into the related branches I’m working on to keep everything aligned. Thanks again for all your hard work and help! 😊

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -30,8 +31,8 @@
- * @author      Jonas Sajonas
- * @license     https://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ * @author        Jonas Sajonas
+ * @license        https://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
```
Not sure if this change was something automatic by your editor but the existing indentation was correct 😄

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -59,13 +60,14 @@ public function __construct(int $facetofaceid) {
+     * @throws moodle_exception If the file cannot be loaded.
      */
-    public function load_from_file(int $fileid) {
+    public function load_from_file(int $fileid):bool {
```
There should be a space after the colon for the return type here. e.g. `...): bool {`

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -88,7 +90,7 @@ public function load_from_file(int $fileid) {
      * @return array An indexed array of column headers.
      */
-    public static function get_headers() {
+    public static function get_headers():array {
```
Return type spacing should be corrected here too

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -127,56 +128,48 @@ private function get_iterator(): Generator {
-
-        foreach ($headerline as &$field) {
-            $field = trim($field, "'");
+            throw new moodle_exception(
+                'error:bookingsuploadfileheaderfieldmismatch',
+                'mod_facetoface',
+                get_string('error:noheaderrow', 'facetoface')
+            );
```
Looks like the previous comment on this one wasn't addressed:
> I think reusing 'error:bookingsuploadfileheaderfieldmismatch' here isn't really the best move since it specifically mentions a "Mismatched number of fields" but that is not the case.
You're also running into a case of effectively concatenating language strings by passing a language string into the $a parameter. Although in this case it's actually being passed to the $link param which I suspect wasn't the intention.
> 
> You should instead just implement the 'error:noheaderrow' string and use that directly as the exception's $errorcode.

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -127,56 +128,48 @@ private function get_iterator(): Generator {
-
-                foreach ($data as &$field) {
-                    $field = trim($field, "'");
+                    throw new moodle_exception(
+                        'error:bookingsuploadfileheaderfieldmismatch',
+                        'facetoface'
+                    );
```
This one still needs an incrementing value to count the row number since that language string has "... on row {$a}."

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -336,7 +326,7 @@ public function validate(): array {
      * @return bool True if all sessions were created successfully, false otherwise.
      */
-    public function process() {
+    public function process():bool {
```
This should have return type spacing updated

NickLambellGCH replied:
A bit of a secondary comment from an existing bit of feedback that it looks like might have been missed is that all the calls like `$this->errors[] = ...` in this function are directly adding strings to the errors array where they should be adding `[ index, string ]` like in `validate()`.

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -450,20 +483,21 @@ public function process() {
      * @return array A list of error entries.
      */
-    public function get_errors() {
+    public function get_errors():array {
```
Return type spacing to be corrected here

NickLambellGCH commented on classes/bulk_session_manager.php:
```diff
@@ -450,20 +483,21 @@ public function process() {
+     * @return array List of CSV records.
      */
-    public function get_records() {
+    public function get_records():array {
```
Return type spacing to be corrected here

NickLambellGCH replied:
Looks like this one was missed 😄

NickLambellGCH commented on classes/event/csv_processed_bulksession.php:
```diff
@@ -15,23 +15,27 @@
  * @copyright   2025 Gold Coast Health
- * @author      Jonas Sajonas
- * @license     https://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ * @author        Jonas Sajonas
+ * @license        https://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
```
Looks like the indentation god adjusted here too

NickLambellGCH commented on classes/event/csv_processed_bulksession.php:
```diff
@@ -15,23 +15,27 @@
+ * @license        https://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
  */
-class csv_processed_bulksession extends \core\event\base {
+
```
Shouldn't need this blank line since the PHPDoc is directly for the class

NickLambellGCH commented on classes/event/csv_processed_bulksession.php:
Most of the functions in this class will need the return type spacing corrected 😄

NickLambellGCH commented on classes/form/bulk_session_confirm_form.php:
```diff
@@ -28,8 +28,8 @@
  * @copyright   2025 Gold Coast Health
- * @author      Jonas Sajonas
- * @license     https://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ * @author        Jonas Sajonas
+ * @license        https://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
```
Indentation here too

NickLambellGCH commented on classes/form/bulk_session_upload_form.php:
```diff
@@ -31,11 +31,10 @@
  * @copyright   2025 Gold Coast Health
- * @author      Jonas Sajonas
- * @license     https://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ * @author        Jonas Sajonas
+ * @license        https://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
```
This indentation too 😄

NickLambellGCH commented on lang/en/facetoface.php:
```diff
@@ -867,67 +870,56 @@
 $string['facetoface:validatebulksessions'] = 'Validate Bulk Sessions Upload';
-$string['error:bulkuploaderrors'] = 'Errors were found in the bulk upload file.';
+$string['error:uploadsessionserrorsfound'] = 'Errors were found in the uploaded file. Sessions cannot be processed until they are resolved.';
+$string['facetoface:csvline'] = 'CSV line';
+$string['facetoface:norecordsfound'] = 'No records found';
+$string['facetoface:bulksessionsprocessed'] = 'Bulk sessions have been successfully processed.';
```
You typically shouldn't need to prefix language strings with the plugin name since this is in its own distinct language file and is always referenced with the plugin name in `get_string()`

NickLambellGCH commented on uploadbulksessions.php:
```diff
@@ -20,8 +20,8 @@
  * @copyright   2025 Gold Coast Health
- * @author      Jonas Sajonas
- * @license     https://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ * @author        Jonas Sajonas
+ * @license        https://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
```
This indentation too

NickLambellGCH commented on uploadbulksessions.php:
```diff
@@ -64,25 +65,37 @@
+ * @param array $errors An array of errors to display.
+ * @return void
+ */
+function handle_bulk_upload_errors($errors):void {
```
Return type spacing on this one will need to be tweaked too

NickLambellGCH commented on uploadbulksessions.php:
```diff
@@ -64,25 +65,37 @@
     echo $OUTPUT->header();
-    echo $OUTPUT->notification(get_string('error:bulkuploadfileerrorsfound', 'mod_facetoface', count($errors)), 'error');
+    echo $OUTPUT->notification(
+        get_string('error:uploadsessionserrorsfound', 'mod_facetoface', count($errors)),
```
Looks like this language string isn't using the error count that's passed in. Might as well update the language string to be like 'error:bookingsuploadfileerrorsfound' and contain the error count.

NickLambellGCH reviewed (changes requested):
Thanks for all the changes! Just a few more feedback items on this one 😄

NickLambellGCH commented on lang/en/facetoface.php:
```diff
@@ -889,12 +889,11 @@
 $string['error:couldnotsavecustomfieldshort'] = 'Could not save custom field with short name "{$a}"';
-$string['facetoface:uploadbulksessions'] = 'Upload Bulk Sessions';
-$string['facetoface:uploadandpreviewbulk'] = 'Upload & Preview Bulk Sessions';
+$string['uploadbulksessions'] = 'Upload Bulk Sessions';
```
While this change was correct since the string is used for page output, it looks like it was also pulling double duty as it also acted as the label for the 'mod/facetoface:uploadbulksessions' capability.
To resolve this, you'll want to add another language string with the original 'facetoface:uploadbulksessions' identifier as well 😄

NickLambellGCH commented:
@Sajonasj-OC I also just noticed that you're conflicting with MOODLE_DEV_LOL so you'll want to merge that in and resolve conflicts

YuchengZhuQH commented on db/access.php:
```diff
@@ -217,4 +217,16 @@
+        'contextlevel' => CONTEXT_COURSE,
+        'archetypes' => [
+            'teacher' => CAP_ALLOW,
+            'editingteacher' => CAP_ALLOW,
```
`'coursecreator'` has the right to `'mod/facetoface:editsessions'`. They can add, edit, copy and delete a session. We may also want to give them the permission to bulk-upload sessions.

YuchengZhuQH commented on example_bulk.csv:
```diff
@@ -0,0 +1,6 @@
+yes,15/04/2025,14:00,15/04/2025,16:30,no,50,yes,150,200,180,Leadership Workshop,venue,Conference Hall
+yes,20/06/2025,10:30,20/06/2025,12:30,yes,30,no,120,150,120,Technical session,trainer,John Doe
+yes,10/07/2025,08:00,10/07/2025,10:00,yes,40,no,120,90,70,New employee onboarding,department,HR
+yes,15/08/2025,11:00,15/08/2025,13:00,no,25,yes,120,120,90,Cybersecurity Awareness,location,Room B1
```
Please check if the example file is correct in Excel. There are 13 fields in the header, but 14 in the body.

Sajonasj-OC replied:
I added the optional customfields for clarity, updated the lang string and added it to the upload form instructions.

YuchengZhuQH commented on lang/en/facetoface.php:
```diff
@@ -887,3 +890,61 @@
+$string['error:failedtocreatesession'] = 'Failed to create session';
+$string['error:failedtocreatedates'] = 'Failed to create dates for session #{$a}';
+$string['error:couldnotsavecustomfieldshort'] = 'Could not save custom field with short name "{$a}"';
+$string['facetoface:uploadbulksessions'] = 'Upload Bulk Sessions';
```
A small issue that could make the UI look more professional:
Currently, the UI looks like
> Add a new session
> 
> Upload Bulk Sessions
> 
> Upload bookings

One problem is the case inconsistency. The other is that Moodle users are familiar with "Import..." for the idea of bulk uploading using a CSV file, so that word needs to be there.

We might change it to something like "Import multiple sessions".

YuchengZhuQH replied:
Or maybe "Upload multiple sessions" to be consistent with "Upload bookings" below?

YuchengZhuQH commented on classes/form/bulk_session_upload_form.php:
```diff
@@ -0,0 +1,209 @@
+
+        // Example CSV link.
+        $url = new moodle_url('/mod/facetoface/example_bulk.csv');
+        $link = html_writer::link($url, get_string('examplecsv', 'mod_facetoface'));
```
The file downloaded is called "example_bulk.csv", but the string on the UI is `example.csv`. Also, there is another bulk uploading link immediately below, causing confusion.
We might change the string to something like "import_multiple_sessions_example.csv" in all places. Probably using only one string in `lang/en/facetoface`.

YuchengZhuQH replied:
Or maybe "upload_multiple_sessions_example.csv" to be consistent with"Upload bookings" below?

Sajonasj-OC replied:
All the example files across Moodle are named example.csv, so we followed the same convention for familiarity. While the link name remains example.csv for consistency, the actual downloaded file uses a more descriptive name.

Would you like me to change it, or are you happy to keep it?

YuchengZhuQH commented on classes/form/bulk_session_upload_form.php:
```diff
@@ -0,0 +1,209 @@
+        $mform->addElement(
+            'static',
+            'example_bulkcsv',
+            get_string('examplecsv', 'mod_facetoface'),
```
Same as above.

YuchengZhuQH commented on example_sessions.csv:
```diff
@@ -0,0 +1,2 @@
+Session Date/Time Known,Start Date,Start Time,Finish Date,Finish Time,Allow Cancellations,Capacity,Allow Overbookings,Duration,Normal Cost,Discount Cost,Details,Customfield_Facility,Customfield_Location,Customfield_Room
```
`Customfield_Facility,Customfield_Location,Customfield_Room` needs to be lower case, like `Customfield_facility,Customfield_location,Customfield_room`, because the `shortname`s on GCHLOL are in lower case.

YuchengZhuQH replied:
Potentially helpful to tell the user the names are not defined.

YuchengZhuQH commented on uploadbulksessions.php:
```diff
@@ -0,0 +1,237 @@
+            continue;
+        }
+
+        $line = $error[0];
```
The error table looks great!
There's a small confusion, though - When "CSV line" shows 2, the user actually sees line 4 in the Excel file.
I'm thinking maybe adjust the row number by +2 can reduce confusion:
```
        $excelrowoffset = 2;  // Ensure the line number is the same as in the Excel file.
        $line = $error[0] + $excelrowoffset;
```

YuchengZhuQH commented on lang/en/facetoface.php:
```diff
@@ -887,3 +890,62 @@
+$string['error:failedtocreatesession'] = 'Failed to create session';
+$string['error:failedtocreatedates'] = 'Failed to create dates for session #{$a}';
+$string['error:couldnotsavecustomfieldshort'] = 'Could not save custom field with short name "{$a}"';
+$string['facetoface:uploadbulksessions'] = 'Upload Bulk Sessions';
```
Just nitpicking: 'Upload Bulk Sessions' sounds like each session has a lot of users. It should be 'Upload Sessions in Bulk" (More similar to "Upload" on the "Upload bookings" page) or "Bulk Upload Sessions". Just be consistent everywhere.

YuchengZhuQH commented on lang/en/facetoface.php:
```diff
@@ -887,3 +890,62 @@
+$string['error:failedtocreatedates'] = 'Failed to create dates for session #{$a}';
+$string['error:couldnotsavecustomfieldshort'] = 'Could not save custom field with short name "{$a}"';
+$string['facetoface:uploadbulksessions'] = 'Upload Bulk Sessions';
+$string['uploadbulksessions'] = 'Upload multiple sessions';
```
That was my bad. I just realised that the link below is "Upload bookings". Probably better to call this one "Upload sessions" to be consistent. Currently, it looks like:
```
Add a new session

Upload multiple sessions

Upload bookings
```

YuchengZhuQH commented on lang/en/facetoface.php:
```diff
@@ -887,3 +890,62 @@
+$string['error:couldnotsavecustomfieldshort'] = 'Could not save custom field with short name "{$a}"';
+$string['facetoface:uploadbulksessions'] = 'Upload Bulk Sessions';
+$string['uploadbulksessions'] = 'Upload multiple sessions';
+$string['uploadandpreviewbulk'] = 'Upload & Preview Bulk Sessions';
```
"Upload and preview sessions" would be consistent with "Upload and validate bookings".
